### PR TITLE
 Stop using ActiveXML in Buildresult 

### PR DESCRIPTION
**Pull request by coolo**
I'll rebase once the 1st commit is in (which is already approved), so please concentrate on the Buildresult class/module

### CODE REVIEW

File: src/api/app/models/project.rb [hunk 1]
```diff
@@ -1419,24 +1419,24 @@ def build_succeeded?(repository = nil)
...
 
-    br = Buildresult.find(project: name, view: 'summary')
+    br = Buildresult.find_hashed(project: name, view: 'summary')
     # no longer there?
-    return false unless br
+    return false if br.empty?
 
-    br.each('result') do |result|
-      if repository && result.value(:repository) == repository
+    br.elements('result') do |result|
```
Comment (bgeuken) [inline src/api/app/models/project.rb]: There isn't an `elements` method if this isn't ActiveXML, or? 🤔

Comment (coolo) [reply src/api/app/models/project.rb]: elements is a Xmlhash function

Comment (bgeuken) [reply src/api/app/models/project.rb]: Ah, right. Forgot that we don't get a "normal" hash here.

File: src/api/app/models/buildresult.rb [hunk 2]
```diff
@@ -39,6 +39,16 @@ class Buildresult < ActiveXML::Node
     unknown:      'The scheduler has not yet evaluated this package. Should be a short intermediate state for new packages.'
   }.with_indifferent_access.freeze
 
+  def self.find_hashed(opts = {})
```
Comment (mdeniz) [inline src/api/app/models/buildresult.rb]: [Nitpick]

This is something that to me it feels that belongs to that second level (not implemented yet) in the `Backend::*` library, that one that reads from the backend a string, then parses it with one XML library and returns a basic type for user's consumption.

File: src/api/lib/backend/api/build_results/status.rb [hunk 3]
```diff
@@ -24,9 +24,12 @@ def self.build_reason(project_name, package_name, repository_name, architecture_
 
         # Returns the result view for a build
         # @return [String]
-        def self.build_result(project_name, package_name, repository_name, architecture_name)
-          http_get(['/build/:project/_result', project_name],
-                   params: { view: :status, package: package_name, arch: architecture_name, repository: repository_name })
+        def self.build_result(project_name, params = {})
```
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/lib/backend/api/build_results/status.rb]: @mdeniz 

> I wouldn't change it to have a **'generic'** params param (that can have package or not... for instance). I rather prefer to stick to mandatory params as we have `(project_name,` `package_name,` `repository_name,` `architecture_name`) plus the params you want to add like `code` in this case.

I guess Rubocop won't like this... :unamused:

[the remark this answers is not in the document]
Comment (mdeniz) [reply src/api/lib/backend/api/build_results/status.rb]: Then change the cop

[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/lib/backend/api/build_results/status.rb]: @mdeniz 

> Then change the cop

no... this is a cop we already agreed on... if you don't like it our documentation say it should be discussed outside the PR... no style discussions in PR :bear:

[the remark this answers is not in the document]
Comment (mdeniz) [reply src/api/lib/backend/api/build_results/status.rb]: For sure, this is not about Rubocop or changing rules... that could be a side effect.

The main thing is to use verbose params and a way of requiring them...

[the remark this answers is not in the document]
Comment (coolo) [reply src/api/lib/backend/api/build_results/status.rb]: I understand you don't want swiss knife functions, but just look at how project::monitor uses this route (indirectly for now). There is no mandatory field in there :(

**Comment by coolo** [comment]
The scariest part of this PR is this:
screencapture-circleci-workflow-run-d0e65b5e-836e-4d5b-9bdc-63a7dfc49c16-2018-08-06-18_29_48

File: src/api/lib/backend/api/build_results/status.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (mdeniz) [reply src/api/lib/backend/api/build_results/status.rb]: That's what we want to avoid :+1:

**Comment by mdeniz** [comment]
@coolo my request of changes wasn't intended to revert yours in the **lib/backend**, it was kind of ok what you did, I only had asked to change the way the params are defined. 

But what I clearly don't like at all is introducing the code that calculates the url based on the params inside the model, that is already in the backend library and is tested. And BTW your PR says stop using `ActiveXML` and now in the line 55 of the `find_xml` you still use it! :smiling_imp: 

So, please recover what you did when using the backend library and we can talk about the best way of doing it :heart:

**Comment by coolo** [comment]
I'm mostly interested in killing ActiveXML::Node first actually. 

But please have a look at #5570 and I'll revist this one later

**Comment by mdeniz** [comment]
In #5570 you got it. That's the intention behind the library, that you call a method and it is transformed in the right thing in the backend.

Please, do the same in this PR.

**Comment by coolo** [comment]
I reverted the revert, but named my backend function result_swiss_knife and sorted out 2 use cases. Here are the remaining users

```
app/controllers/webui/package_controller.rb:    results_from_backend = Buildresult.find_hashed(project: @project, package: @package_name, repository: @repository, view: ['binarylist', 'status'])
app/controllers/webui/package_controller.rb:    @buildresult = Buildresult.find_hashed(project: @project.to_param, package: @package.to_param, view: 'status')
app/controllers/webui/project_controller.rb:    @buildresult = Buildresult.find_hashed(find_opt)
app/controllers/webui/project_controller.rb:      @buildresult = Buildresult.find_hashed(project: params[:project], package: params[:package], view: 'status', lastbuild: 1)
app/models/kiwi/image.rb:      results = ::Buildresult.find_hashed(project: package.project, package: package.name, view: 'status', multibuild: '1', locallink: '1')
app/models/local_build_result/for_package.rb:      backend_results = Buildresult.find_hashed(project: project, package: package, view: 'status', multibuild: '1', locallink: '1')
app/models/obs_factory/distribution_strategy_casp.rb:      buildresult = Buildresult.find_hashed(project: project.name, package: "CAASP-dvd5-DVD-#{arch}", repository: 'images', view: 'binarylist')
app/models/obs_factory/distribution_strategy_factory.rb:      buildresult = Buildresult.find_hashed(project: project.name, package: "#{test_dvd_prefix}-#{arch}", repository: 'images', view: 'binarylist')
app/models/obs_factory/obs_project.rb:        ::Buildresult.find_hashed(project: name, view: 'summary')
app/models/obs_factory/staging_project.rb:          buildresult = Buildresult.find_hashed(project: name, view: 'summary', repository: current_repo['repository'], arch: current_repo['arch'])
app/models/project.rb:    br = Buildresult.find_hashed(project: name, view: 'summary')

**Comment by coolo** [comment]
@mdeniz I would really prefer to leave it like this. I don't intend to refactor all backend calls - I want to get rid of ActiveXML::Node. We can kill the find_hashed calls later

**Comment by coolo** [comment]
@Ana06's 'requested changes' was for the stuff I reverted meanwhile.

**Comment by vpereira** [comment]
@coolo build failed on BS. Maybe just a rebuild will fix it.

**Pull request closed without merging**